handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java

```diff
@@ -0,0 +1,150 @@
+     * @param rules            {@link IpSubnetFilterRule} as array
+     */
+    public IpSubnetFilter(boolean acceptIfNotFound, IpSubnetFilterRule... rules) {
+        this(acceptIfNotFound, Arrays.asList(ObjectUtil.checkNotNull(rules, "rules")));
+    }
+
+    /**
+     * <p> Create new {@link IpSubnetFilter} Instance with specified {@link IpSubnetFilterRule} as {@link List}. </p>
+     * <p> {@code acceptIfNotFound} is set to {@code true} </p>
+     *
+     * @param rules {@link IpSubnetFilterRule} as {@link List}
+     */
+    public IpSubnetFilter(List<IpSubnetFilterRule> rules) {
+        this(true, rules);
+    }
+
+    /**
+     * <p> Create new {@link IpSubnetFilter} Instance with specified {@link IpSubnetFilterRule} as {@link List}
+     * and specify if we'll accept a connection if we don't find it in the rule(s). </p>
+     *
+     * @param acceptIfNotFound {@code true} if we'll accept connection if not found in rule(s).
+     * @param rules            {@link IpSubnetFilterRule} as {@link List}
+     */
+    public IpSubnetFilter(boolean acceptIfNotFound, List<IpSubnetFilterRule> rules) {
+        this.rules = ObjectUtil.checkNotNull(rules, "rules");
+        this.acceptIfNotFound = acceptIfNotFound;
+
+        // Iterate over rules and check for `null` rule.
+        for (IpSubnetFilterRule ipSubnetFilterRule : this.rules) {
+            ObjectUtil.checkNotNull(ipSubnetFilterRule, "rule");
+        }
+
+        sortAndFilter();
+    }
+
+    @Override
+    protected boolean accept(ChannelHandlerContext ctx, InetSocketAddress remoteAddress) {
+        int indexOf = Collections.binarySearch(this.rules, remoteAddress, IpSubnetFilterRuleComparator.INSTANCE);
+        if (indexOf >= 0) {
+            return this.rules.get(indexOf).ruleType() == IpFilterRuleType.ACCEPT;
+        }
+        return acceptIfNotFound;
+    }
+
+    /**
+     * <ol>
+     *     <li> Sort the list </li>
+     *     <li> Remove over-lapping subnet </li>
+     *     <li> Sort the list again </li>
+     * </ol>
+     */
+    private void sortAndFilter() {
+        Collections.sort(this.rules);
+        Iterator<IpSubnetFilterRule> iterator = rules.iterator();
+        List<IpSubnetFilterRule> toRemove = new ArrayList<IpSubnetFilterRule>();
+
+        try {
+            IpSubnetFilterRule parentRule = null;
+            while (iterator.hasNext()) {
+
+                // If parentRule is null, take first element out of Iterator.
+                if (parentRule == null) {
+                    parentRule = iterator.next();
+                }
+
+                // Take one more element out of Iterator, this will be childRule.
+                IpSubnetFilterRule childRule = iterator.next();
+
+                // If parentRule matches childRule, schedule that Rule for deletion.
+                if (parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
+                    toRemove.add(childRule);
```

Comment:
Instead of collecting a list of things to remove (or removing via `Iterator.remove()`), you can instead build up a new list of things you want to keep. This algorithm would be linear in time, instead of N*M from `removeAll`. You can then do the null-checking at the same time. Make this method take the list-to-be-filtered as a parameter, and return the new filtered list. Then you can do all of this at the `this.rules` field assignment.

Reply:
I added code that checks if all rules belong to the same type while checking for `null` rules inside the user-supplied list of rules. If I try to move these inside sorting codes, it'll be a complete mess. 

The reason behind adding this: We don't have to call `List#get(int)` because we already know if the list of rule contains the same `IpFilterRuleType`.

PTAL at the latest code and revert back. :D

Reply:
I mean something like this:

```patch
Index: handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java	(revision 295e104c7ed9027584927777073ae9f2f81cc2df)
+++ handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java	(date 1598435434667)
@@ -86,14 +86,14 @@
      * @param rules            {@link IpSubnetFilterRule} as {@link List}
      */
     public IpSubnetFilter(boolean acceptIfNotFound, List<IpSubnetFilterRule> rules) {
-        this.rules = ObjectUtil.checkNotNull(rules, "rules");
+        ObjectUtil.checkNotNull(rules, "rules");
         this.acceptIfNotFound = acceptIfNotFound;
 
         int numAccept = 0;
         int numReject = 0;
 
         // Iterate over rules and check for `null` rule.
-        for (IpSubnetFilterRule ipSubnetFilterRule : this.rules) {
+        for (IpSubnetFilterRule ipSubnetFilterRule : rules) {
             ObjectUtil.checkNotNull(ipSubnetFilterRule, "rule");
             if (ipSubnetFilterRule.ruleType() == IpFilterRuleType.ACCEPT) {
                 numAccept++;
@@ -120,7 +120,7 @@
             ipFilterRuleType = null;
         }
 
-        sortAndFilter();
+        this.rules = sortAndFilter(rules);
     }
 
     @Override
@@ -143,39 +143,27 @@
      *     <li> Sort the list again </li>
      * </ol>
      */
-    private void sortAndFilter() {
-        Collections.sort(this.rules);
+    private static List<IpSubnetFilterRule> sortAndFilter(List<IpSubnetFilterRule> rules) {
+        Collections.sort(rules);
         Iterator<IpSubnetFilterRule> iterator = rules.iterator();
-        List<IpSubnetFilterRule> toRemove = new ArrayList<IpSubnetFilterRule>();
+        List<IpSubnetFilterRule> toKeep = new ArrayList<IpSubnetFilterRule>();
 
-        IpSubnetFilterRule parentRule = null;
+        IpSubnetFilterRule parentRule = iterator.hasNext() ? iterator.next() : null;
+        if (parentRule != null) {
+            toKeep.add(parentRule);
+        }
         while (iterator.hasNext()) {
-
-            // If parentRule is null, take first element out of Iterator.
-            if (parentRule == null) {
-                parentRule = iterator.next();
-            }
-
-            // If we don't have any more rule, we're done now.
-            if (!iterator.hasNext()) {
-                break;
-            }
-
-            // Take one more element out of Iterator, this will be childRule.
+            // Grab a potential child rule.
             IpSubnetFilterRule childRule = iterator.next();
 
-            // If parentRule matches childRule, schedule that Rule for deletion.
-            if (parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
-                toRemove.add(childRule);
-            } else {
-                // If parentRule does not matches childRule, this childRule will become new parentRule
-                // and we'll do the same again
+            // If parentRule matches childRule, then there's no need to keep the child rule.
+            // Otherwise, the rules are distinct and we need both.
+            if (!parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
+                toKeep.add(childRule);
+                // Then we'll keep the child rule around as the parent for the next round.
                 parentRule = childRule;
             }
         }
-
-        rules.removeAll(toRemove);
-        toRemove.clear();
-        Collections.sort(rules); // Re-sort just to be sure
+        return toKeep;
     }
 }
```

Reply:
@hyperxpro Not sure if I have access to directly push to this PR, but I can give it a go if you like.

Reply:
Just create a gist of updated source and I will copy from there. :p

Reply:
If you use Intellij you can also copy the snippet above and click VCS→Apply Patch from Clipboard…

Reply:
NVM, Did it myself. Thanks for helping a noob. :p